added a helper to convert certain GAP matrices

ThomasBreuer opened the pull request:
Stevell Muller has asked in Oscar Slack for such functionality, as far as I understand.
Does this function what is needed?
(An obvious extension would be to enter the intended cyclotomic field as an argument.)

ThomasBreuer committed:
added a helper to convert certain GAP matrices

Stevell Muller has asked in Oscar Slack for such functionality, as far as I understand.

fingolfin commented:
@StevellM perhaps take a look, this might help you?

fingolfin commented on src/GAP/gap_to_oscar.jl:
```diff
@@ -43,3 +43,29 @@ function GAP.gap_to_julia(::Type{fmpq_mat}, obj::GAP.GapObj)
+      n = GAP.Globals.NumberColumns(mat)
+      entries = [F(Vector{fmpz}(GAP.Globals.Coefficients(B, mat[i, j])))
+                 for i in 1:m for j in 1:n]
+      push!(result, matrix(F, m, n, entries))
```
If you do it like this, then `entries` is an `m x n` matrix and you can simplify the next line
```suggestion
      entries = [F(Vector{fmpz}(GAP.Globals.Coefficients(B, mat[i, j])))
                 for i in 1:m, j in 1:n]
      push!(result, matrix(F, entries))
```

fingolfin commented on src/GAP/gap_to_oscar.jl:
```diff
@@ -43,3 +43,29 @@ function GAP.gap_to_julia(::Type{fmpq_mat}, obj::GAP.GapObj)
+    root = GAP.Globals.E(N)
+    elms = [root^i for i in 0:(euler_phi(N)-1)]
+    B = GAP.Globals.Basis(gapF, GAP.GapObj(elms))
+    F, z = CyclotomicField(N)
```
How about also adding a method which enables one to write `F(x)` where F is a cyclotomic field and `x` a GAP cyclotomic? I.e., something like this:
```julia
function (a::AnticNumberField)(x::GapObj)
   GAP.Globals.IsCyclotomic(b) || throw(ArgumentError("input is not a GAP cyclotomic"))
   B = [... compute basis like above... perhaps we could also cache these in a `WeakValueDict` from AbstractAlgebra ]
  return F(Vector{fmpz}(GAP.Globals.Coefficients(B, x)))
end
```

Granted, we probably wouldn't want to use it in `matrices_over_cyclotomic_field` for performance reasons (recomputing the basis again and again is bad -- unless we cache it...), but it might still be useful for some users, no?

Hmm, and I guess taking this one step further, perhaps a `matrix(F, mat)` (where `mat` is the GAP matrix) could/should be made to work?

fingolfin commented on src/GAP/gap_to_oscar.jl:
```diff
@@ -43,3 +43,29 @@ function GAP.gap_to_julia(::Type{fmpq_mat}, obj::GAP.GapObj)
+      push!(result, matrix(F, m, n, entries))
+    end
+
+    return convert(Vector{typeof(result[1])}, result), F, z
```
Aha, that's of course to work around the type instability caused by doing `result = []` above.

I think one could do this above instead: `result = elem_type(MatrixSpace(F, 1, 1))[]`

thofma replied:
`dense_matrix_type(F)[]`.

PS: If you really really need to do this somewhere, just use `identity.(result)`. (EDIT: Does not apply here)

ThomasBreuer replied:
Thanks @thofma, this is much better.

ThomasBreuer replied:
Concerning the hint about `identity.(result)` (which is not needed here):
To me, this looks like relying on a (documented?) side-effect, whereas the call `convert(Vector{T}, result)` expresses exactly what I want to achieve. Moreover, if I know the intended type `T` in question, the `convert` call will signal an error if some entry of `result` does not fit to `T`, but `identity.(result)` will not complain.

thofma replied:
Yes, you are right. It is not what you want in this application.

ThomasBreuer committed:
extended the code according to the suggestions

ThomasBreuer committed:
provide also conversions in the other direction

convert elements of cyclotomic fields in Oscar, and matrices over these fields, to GAP

fingolfin commented on src/GAP/gap_to_oscar.jl:
```diff
@@ -38,8 +38,66 @@ fmpz_mat(obj::GAP.GAP.GapObj) = gap_to_julia(fmpz_mat, obj)
+function (F::AnticNumberField)(obj::GAP.GapObj)
+    Nemo.iscyclo_type(F) || throw(ArgumentError("F is not a cyclotomic field"))
+    GAP.Globals.IsCyclotomic(obj) || throw(ArgumentError("input is not a GAP cyclotomic"))
+    N = Oscar.get_special(F, :cyclo)
```
Perhaps there could be a `conductor` method returning this value? @fieker ?

fingolfin commented on src/GAP/gap_to_oscar.jl:
```diff
@@ -38,8 +38,66 @@ fmpz_mat(obj::GAP.GAP.GapObj) = gap_to_julia(fmpz_mat, obj)
+    B = default_basis_GAP_cyclotomic_field(N)
+    m = GAP.Globals.NumberRows(mat)
+    n = GAP.Globals.NumberColumns(mat)
+    entries = [F(Vector{fmpz}(GAP.Globals.Coefficients(B, mat[i, j])))
+               for i in 1:m, j in 1:n]
```
I guess this will produce an error if `F` is too small to represent all entries of `mat`. Perhaps you could add a test case for that situation, just to make sure it really produces an error and not garbage.

ThomasBreuer replied:
There is already such a test, the attempt to convert a matrix containing `E(5)` to a matrix over `CyclotomicField(7)`.
(`GAP.Globals.Coefficients` returns `fail` in this case.)

fingolfin commented on test/GAP/gap_to_oscar.jl:
```diff
@@ -111,3 +111,61 @@ end
+    @test F(GAP.evalstr("2^64")) == F(2)^64
+
+    F, z = CyclotomicField(5)
+    @test F(GAP.evalstr("Sqrt(5)")) == -2*z^3 - 2*z^2 - 1
```
```suggestion
    @test F(GAP.Globals.Sqrt(5)) == -2*z^3 - 2*z^2 - 1
```

fingolfin commented on test/GAP/gap_to_oscar.jl:
```diff
@@ -111,3 +111,61 @@ end
+
+    F, z = CyclotomicField(15)
+    @test F(GAP.evalstr("E(5)")) == z^3
+    @test F(GAP.evalstr("E(3)")) == z^5
```
```suggestion
    @test F(GAP.Globals.E(5)) == z^3
    @test F(GAP.Globals.E(3)) == z^5
```

and so one elsewhere in this file

fingolfin commented on test/GAP/oscar_to_gap.jl:
```diff
@@ -78,6 +78,22 @@ end
 
+@testset "single cyclotomics" begin
+    F, z = CyclotomicField(5)
+    @test GAP.GapObj(z^2+z+1) == GAP.evalstr("E(5)^2+E(5)+1")
```
```suggestion
     e5 = GAP.Globals.E(5)
    @test GAP.GapObj(z^2+z+1) == e5^2+ e5 +1
```

ThomasBreuer committed:
beautified some tests

ThomasBreuer commented:
Again (see #718): no tests are running?
